File: 0300-longest-increasing-subsequence/0300-longest-increasing-subsequence.py
class Solution:
    def lengthOfLIS(self, nums: List[int]) -> int:
        # DP - https://www.youtube.com/watch?v=mouCn3CFpgg
        n = len(nums)
        dp = [1]*(n)
        
        
        for i in range(n):
            for j in range(i):
                if nums[i] > nums[j] and dp[i] <= dp[j]:
                    dp[i] = 1 + dp[j]
        
        return max(dp)
    
    
        mono = []
        for num in nums:
            i = bisect.bisect_left(mono, num)  # or bisect.bisect_right
            if i == len(mono):
                mono.append(num)
            else:
                mono[i] = num
        print(mono)
        return len(mono)

File: 0300-longest-increasing-subsequence/0300-longest-increasing-subsequence.java
class Solution {
    public int lengthOfLIS(int[] nums) {
        int n = nums.length;

        List<Integer> tails = new ArrayList<>(); 
        int[] tailsIndex = new int[n];   

        int[] parent = new int[n];  // for reconstruction
        Arrays.fill(parent, -1);

        for (int i = 0; i < n; i++) {
            int pos = Collections.binarySearch(tails, nums[i]); // bisect_left
            if (pos < 0) pos = -(pos + 1);  // insertion point

            if (pos == tails.size()) tails.add(nums[i]);
            else tails.set(pos, nums[i]);

            tailsIndex[pos] = i;

            // if there is prev element
            if (pos > 0)    parent[i] = tailsIndex[pos - 1];
        }

        // Reconstruct LIS
        List<Integer> result = new ArrayList<>();
        int index = tailsIndex[tails.size() - 1];

        while (index != -1) {
            result.add(nums[index]);
            index = parent[index];
        }

        Collections.reverse(result);
        // System.out.println("Longest Icreasing Subsequence : " + result);
        return result.size();   // return tails.size();
    }
}